### PR TITLE
Improve step time by 5% by avoiding useless all-gather on logits

### PR DESCRIPTION
**Pull request by py4**
# Description

Returning full unsharded logits is costly and results in expensive all-gather. It's ~6ms for 2K tokens with 128K vocab.
        # In future, if we need returning logits, it should be through topK not the entire logits, or at least through a flag not a default.

# Tests

local tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by wwl2755-google** [comment]
LGTM!
